### PR TITLE
fix breaks of the deployment pipeline for on premise machines and updated kernel

### PR DESCRIPTION
**Pull request by YinYangOfDao**

### CODE REVIEW

File: src/ClusterBootstrap/deploy.py [hunk 1]
```diff
@@ -572,6 +572,12 @@ def get_ETCD_master_nodes_from_cluster_portal(clusterId):
 
 def get_ETCD_master_nodes_from_config(clusterId):
     Nodes = get_nodes_from_config("infrastructure")
+    if int(config["etcd_node_num"]) == 1:
+        for nodename in config["machines"]:
+            nodeInfo = config["machines"][nodename]
+            if "role" in nodeInfo and nodeInfo["role"]=="infrastructure":
+                config["etcd_private_ip"] = nodeInfo["private-ip"]
```
Comment (hongzhili) [inline src/ClusterBootstrap/deploy.py]: add assert "private-ip" in nodeInfo

File: src/ClusterBootstrap/scripts/setup_nfs_server.sh [hunk 2]
```diff
@@ -1,10 +1,31 @@
...
 
-echo "/data/share {{cnf["cloud_config"]["vnet_range"]}}(rw,sync,no_subtree_check,no_root_squash)" | sudo tee /etc/exports
-sudo systemctl restart nfs-kernel-server
+echo "/data/share 10.209.247.0/24(rw,fsid=1,nohide,insecure,sync,no_subtree_check,no_root_squash)" | sudo tee -a /etc/exports
+
+echo "/data/share 127.0.0.1/32(rw,fsid=1,nohide,insecure,sync,no_subtree_check,no_root_squash)" | sudo tee -a /etc/exports
+
+echo "/data/share 127.0.1.1/32(rw,fsid=1,nohide,insecure,sync,no_subtree_check,no_root_squash)" | sudo tee -a /etc/exports
+
+echo "/data/share 104.44.112.0/24(rw,fsid=1,nohide,insecure,sync,no_subtree_check,no_root_squash)" | sudo tee -a /etc/exports
```
Comment (hongzhili) [inline src/ClusterBootstrap/scripts/setup_nfs_server.sh]: why hard coding here?

File: src/ClusterBootstrap/template/RestfulAPI/config.yaml [hunk 3]
```diff
@@ -33,10 +33,9 @@ webportal_node: {{cnf["webportal_node"]}}
 datasource : {{cnf["datasource"]}}
 kube_custom_scheduler: {{cnf["kube_custom_scheduler"]}}
 WinbindServers: {{cnf["WinbindServers"]}}
-azure_cluster :
-  worker_node_num : {{cnf["azure_cluster"][cnf["cluster_name"]]["worker_node_num"]}}
-  worker_vm_size : {{cnf["azure_cluster"][cnf["cluster_name"]]["worker_vm_size"]}}
-sku_mapping: {{cnf["sku_mapping"]}}
```
Comment (hongzhili) [inline src/ClusterBootstrap/template/RestfulAPI/config.yaml]: do we still need "sku_mapping"?

File: src/ClusterBootstrap/az_tools.py [hunk 4]
```diff
@@ -718,14 +741,13 @@ def gen_cluster_config(output_file_name, output_file=True, no_az=False):
     cc["deploydockerETCD"] = False
     cc["platform-scripts"] = "ubuntu"
     cc["basic_auth"] = "%s,admin,1000" % uuid.uuid4().hex[:16]
-    domain_mapping = {"regular":"%s.cloudapp.azure.com" % config["azure_cluster"]["azure_location"], "low": config["domain_name"]}
+    domain_mapping = {"regular":"%s.cloudapp.azure.com" % config["azure_cluster"]["azure_location"], "low": "redmond.corp.microsoft.com"}
```
Comment (hongzhili) [inline src/ClusterBootstrap/az_tools.py]: why do we use hard coded "redmond.corp.microsoft.com" ?

File: src/ClusterBootstrap/template/etcd/docker_etcd.sh [hunk 5]
```diff
@@ -6,10 +6,16 @@ docker run -d -v /usr/share/ca-certificates/:/etc/ssl/certs -v /var/etcd:/var/et
...
- -listen-client-urls http://0.0.0.0:{{cnf["etcd3port1"]}} \
- -initial-advertise-peer-urls http://{{cnf["etcd_node_ip"]}}:{{cnf["etcd3portserver"]}} \
- -listen-peer-urls http://0.0.0.0:2380 \
+ {% if cnf["etcd_node_num"] == 1 %}-initial-cluster {{cnf["hostname"]}}=https://{{cnf["etcd_private_ip"]}}:{{cnf["etcd3portserver"]}} \
+ -initial-cluster-state new \
+ -initial-cluster-token {{cnf["clusterId"]}} \
+ -advertise-client-urls https://{{cnf["etcd_private_ip"]}}:{{cnf["etcd3port1"]}} \
+ -listen-client-urls https://{{cnf["etcd_private_ip"]}}:{{cnf["etcd3port1"]}},https://127.0.0.1:{{cnf["etcd3port1"]}} \
+ -initial-advertise-peer-urls https://{{cnf["etcd_private_ip"]}}:{{cnf["etcd3portserver"]}} \
+ -listen-peer-urls https://{{cnf["etcd_private_ip"]}}:{{cnf["etcd3portserver"]}} \
```
Comment (hongzhili) [inline src/ClusterBootstrap/template/etcd/docker_etcd.sh]: do you need to use "0.0.0.0" in "listen-client-urls" and "listen-peer-urls"?

File: src/ClusterBootstrap/scripts/setup_nfs_server.sh (hunk 2 above)
Comment (YinYangOfDao) [reply src/ClusterBootstrap/scripts/setup_nfs_server.sh]: this file is generated. it should not be tracked by git. would add it to gitignore

File: src/ClusterBootstrap/az_tools.py (hunk 4 above)
Comment (YinYangOfDao) [reply src/ClusterBootstrap/az_tools.py]: this is an error happened when I was cherrypicking my changes

File: src/ClusterBootstrap/az_tools.py [hunk 6]
```diff
@@ -489,6 +492,26 @@ def create_cluster(arm_vm_password=None, parallelism=1):
             create_vm_param(i, "nfs", config["azure_cluster"]["nfs_vm_size"], False,
                arm_vm_password, config["azure_cluster"]["nfs_vm"][i] if i < len(config["azure_cluster"]["nfs_vm"]) else None )
 
+def add_workers(arm_vm_password=None, parallelism=1):
+    # assert config["priority"] == "regular" and "vmss cloudinit not supported yet"
+    if config["priority"] == "regular":
+        print("entering")
```
Comment (Anbang-Hu) [inline src/ClusterBootstrap/az_tools.py]: Remove this kind of debugging message

File: src/ClusterBootstrap/template/etcd/docker_etcd.sh (hunk 5 above)
Comment (YinYangOfDao) [reply src/ClusterBootstrap/template/etcd/docker_etcd.sh]: updated

File: src/ClusterBootstrap/utils.py [hunk 7]
```diff
@@ -622,3 +625,40 @@ def mergeDict(configDst, configSrc, bOverwrite):
...
+def mask_num(valid_bit):
+    return int('1'*valid_bit+'0'*(32 - valid_bit), 2)
+
+def remain_num(valid_bit):
+    return int('0'*valid_bit+'1'*(32 - valid_bit), 2)
+
+def keep_widest_subnet(ips):
+    res = set()
+    mask2ip = {}
+    ips = sorted(ips, key = lambda x: x[-2:])
```
Comment (Anbang-Hu) [inline src/ClusterBootstrap/utils.py]: cast to `int`

File: src/ClusterBootstrap/utils.py [hunk 8]
```diff
@@ -622,3 +625,40 @@ def mergeDict(configDst, configSrc, bOverwrite):
...
+        ipv4, mask = ip.split("/")
+        mask = int(mask)
+        ipval = ip2int(ipv4)
+        remnmsk = remain_num(mask)
+        assert (remnmsk & ipval == 0), "invalid ip/mask {}!".format(ip)
+        covered_by_wider_mask = False
+        for msk in mask2ip.keys():
+            # wider mask range
+            if msk < mask:
+                this_masked = ipval & mask_num(msk)
```
Comment (Anbang-Hu) [inline src/ClusterBootstrap/utils.py]: Refactor a bit please

File: src/ClusterBootstrap/template/vmss/vmss.sh.template [hunk 9]
```diff
@@ -0,0 +1,23 @@
...
+    --nsg {{cnf["cluster_name"]}}-nsg \
+    --admin-username {{cnf["admin_username"]}} \
+    --storage-sku Premium_LRS \
+    --data-disk-sizes-gb {{cnf["azure_cluster"]["worker_local_storage_sz"]}} \
+    --data-disk-caching ReadWrite \
+    --ssh-key-values ./deploy/sshkey/id_rsa.pub \
+    --instance-count {{cnf["azure_cluster"]["worker_node_num"]}} \
+    --priority {{cnf["priority"]}} \
+    --disable-overprovision \
+    --eviction-policy Deallocate \
```
Comment (Anbang-Hu) [inline src/ClusterBootstrap/template/vmss/vmss.sh.template]: Parameterize `eviction-policy` and `single-placement-group`

File: src/ClusterBootstrap/az_tools.py [hunk 10]
```diff
@@ -666,7 +686,7 @@ def get_disk_from_vm(vmname):
 def gen_cluster_config(output_file_name, output_file=True, no_az=False):
     if config["priority"] == "low":
         utils.render_template("./template/dns/cname_and_private_ips.sh.template", "scripts/cname_and_ips.sh", config)    
-        utils.exec_cmd_local("chmod +x scripts/cname_and_ips.sh")
+        utils.exec_cmd_local("chmod +x scripts/cname_and_ips.sh;bash scripts/cname_and_ips.sh")
```
Comment (Anbang-Hu) [inline src/ClusterBootstrap/az_tools.py]: add a space after `;`

**Pull request merged**